Modules/Core/Common/include/itkFixedArray.h

```diff
@@ -254,11 +254,19 @@ class ITK_TEMPLATE_EXPORT FixedArray
   bool
   operator==(const FixedArray & r) const;
 
+#if __cplusplus < 202002L
+  /** For C++ versions < C++20 define the != operator.
+   * For C++20 != is now a secondary operator, as such
+   * the compiler will now use a ! applied to the primary
+   * == operator, which makes the implementation redundant
+   * in this context.
+   */
```

Comment:
Very interesting @darrent1974 Which compile error did you get from GCC11 when compiling for C++20? I tried the following on gcc 11.1 `-std=c++20`, at https://godbolt.org/z/P7P9315vb but it does seem to compile fine:

```
class Foo
{
public:
    bool operator==(const Foo&) const { return true; }

    bool operator!=(const Foo& arg) const
    {
        return !(*this == arg);
    }
};

int main()
{
    return (Foo{} == Foo{}) && (Foo{} != Foo{});
}

```

Reply:
Hi @N-Dekker, forgive me for the commit messages. It's my first non-trivial pull-request to ITK and I was a bit uncertain as to how much should be said.

Regarding the changes to itkFixedArray and itkPoint wrt the == and != operators. It was quite subtle and took be a while to nut it out. I was getting "ambiguous overload" errors for the operators on these related classes. See the snippet at the bottom.

After a bit of digging around and googling I came across the following article [https://brevzin.github.io/c++/2019/07/28/comparisons-cpp20/](https://brevzin.github.io/c++/2019/07/28/comparisons-cpp20/) which explained some changes regarding comparison operators.

After digging down into the code and found that the == operators are identical for itkFixedArray and itkPoint which I results in the ambiguity. Given this, I decided (for C++20) to discard the definition of == for itkPoint as its parent is sufficient.

Also, given the changes regarding the now "secondary" != operator it seemed redundant to now define != explicitly on itkFixedArray as the compiler would nicely insert a !==.

Let me know if you want a better commit message.

/ITK/src/Modules/ThirdParty/GoogleTest/src/itkgoogletest/googletest/include/gtest/gtest.h:1527:11: error: ambiguous overload for 'operator==' (operand types are 'const itk::FixedArray<double, 2>' and 'const itk::Point<float, 2>')
 1527 |   if (lhs == rhs) {
      |       ~~~~^~~~~~
In file included from /ITK/src/Modules/Core/Common/include/itkMatrix.h:21,
                 from /ITK/src/Modules/Core/Transform/include/itkMatrixOffsetTransformBase.h:23,
                 from /ITK/src/Modules/Core/Transform/test/itkMatrixOffsetTransformBaseGTest.cxx:20:
/ITK/src/Modules/Core/Common/include/itkPoint.h:127:3: note: candidate: 'bool itk::Point<TCoordRep, NPointDimension>::operator==(const Self&) const [with TCoordRep = float; unsigned int NPointDimension = 2; itk::Point<TCoordRep, NPointDimension>::Self = itk::Point<float, 2>]' (reversed)
  127 |   operator==(const Self & pt) const
      |   ^~~~~~~~
In file included from /ITK/src/Modules/Core/Common/include/itkFixedArray.h:454,
                 from /ITK/src/Modules/Core/Common/include/itkNumericTraits.h:1986,
                 from /ITK/src/Modules/Core/Common/include/itkPoint.h:22,
                 from /ITK/src/Modules/Core/Common/include/itkMatrix.h:21,
                 from /ITK/src/Modules/Core/Transform/include/itkMatrixOffsetTransformBase.h:23,
                 from /ITK/src/Modules/Core/Transform/test/itkMatrixOffsetTransformBaseGTest.cxx:20:
/ITK/src/Modules/Core/Common/include/itkFixedArray.hxx:66:1: note: candidate: 'bool itk::FixedArray<TValue, VLength>::operator==(const itk::FixedArray<TValue, VLength>&) const [with TValue = double; unsigned int VLength = 2]'
   66 | FixedArray<TValue, VLength>::operator==(const FixedArray & r) const
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~

Reply:
> It's my first non-trivial pull-request to ITK and I was a bit uncertain as to how much should be said.

I think your pull-request is quite good already 😃 Now I'm still wondering, would it be necessary to have an `#if __cplusplus < 202002L` for each and every `operator!=`, as long as we still support C++14 and C++17. (We only just dropped C++11!)

Would the C++20 error also occur if  the `operator!=` overloads were defined as `friend` functions (inside the class definitions), instead of member functions? Like this:

    friend bool
    operator!=(const FixedArray & lhs, const FixedArray & rhs)
    {
      return !(lhs == rhs);
    }

In general, it appears preferable to define such operators as `friend` functions instead of member functions anyway.

However, if such `friend` functions can't fix the C++20 errors, another approach could be to add a macro, like `ITK_DEFINE_INEQUAL_OPERATOR_MEMBER_FUNCTION(className)`, which may be defined as "nothing" for C++20. Such a macro could be added to "ITK\Modules\Core\Common\include\itkMacro.h" What do you think?

Reply:
I'm not really sure at the moment 😊

Firstly, I ONLY got the ambiguity error for the itkFixedArray/itkPoint classes for "==" operator, which makes me wonder if the error may have been localized to this particular instance. It seemed happy enough with the existing != operator definition. I got rid of it by choice as I thought it was no longer necessary due to the new handling of !=.

The implication I get from the several articles ([1](https://andreasfertig.blog/2020/06/cpp20-inside-of-a-spaceship/) [2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0515r3.pdf)) I've now read on the new comparison operators is that it appears to negate the need to define "friend" operators due to the new reversible nature of the equality operator. 

So, does that potentially require all != operator definitions to be blanked for C++20 along the lines of a macro as you propose? I'm not familial enough with the ITK codebase overall to know how big a job this may be. 

Also, if my interpretation of the "friend" issue is accurate - could all friend definitions and any potential ordering operators <,> etc also be revisited to simplify for C++20. 

Or, will everything keep working as expected as is, with only the occasional spot fire popping up here and there due to edge cases.